feat: Add UI for configuring SAML in Flagsmith

novakzaballa opened the pull request:
Thanks for submitting a PR! Please check the boxes below:

- [x] I have run [`pre-commit`](https://docs.flagsmith.com/platform/contributing#pre-commit) to check linting
- [x] I have added information to `docs/` if required so people know about the feature!
- [x] I have filled in the "Changes" section below?
- [x] I have filled in the "How did you test this code" section below?
- [x] I have used a [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) title for this Pull Request

## Changes

- Add UI for configuring SAML in Flagsmith wrapped with the `saml_configuration` flag.
## How did you test this code?

- Use an organisation with an enterprise plan
- Go to Organisation Settings -> SAML tab
- Create a SAML configuration

matthewelwell commented:
I've added a new list endpoint in the PR [here](https://github.com/Flagsmith/flagsmith-saml/pull/41) which we will need to update this code to use. It's also worth noting that we allow multiple `SamlConfiguration` objects per organisation, so this PR will need to be updated to handle that case - it seems like it only supports a single configuration at the moment. 

Some other changes that will need to be made: 

 1. We should validate the `name` attribute as it needs to be a valid URL path parameter. This means we should restrict the input to only alphanumeric + `-` and `_`. 
 2. We need to make sure that we handle errors from the API since the name field is unique across all organisations in the platform. 
 3. We should add a tooltip explaining what the name field _is_: "An alphanumeric string to uniquely identify the SAML configuration. You will use this as the 'Organisation name' when logging in via the Flagsmith login page. Must only include alphanumeric characters or hyphens / underscores." 
 4. The SamlConfiguration endpoints use the `name` field as the lookup field whereas this PR seems to be trying to use the numeric `id` field after creation. 

We should also note that (in the near future), we will want to also be able to manage a list of attribute mappings (which consist of a string input, and a select box (values TBC). So we should consider that when designing the UI to allow for multiple SamlConfiguration objects.

matthewelwell commented:
@novakzaballa thanks for the changes, it's looking a lot better now but there are still a few issues: 

 1. The IdP metadata is quite unmanageable in an auto-expanding text input. Ideally we'd have a file upload option here, but I think we'd need a new endpoint for that? 
 2. The bottom of the modal is missing padding (see screenshots below)
 3. It's not obvious that the 'Name' input label has a tooltip, let's add an icon there to make it clear. We should probably add tooltips to the remaining fields (probably with the exception of the IdP metadata which is fairly self explanatory. 
 4. The alignment on the table is quite odd. 
 5. The 'Edit Configuration' button should say 'Update Configuration'
 6. The 'Download Service Provider Metadata' option doesn't work correctly - it downloads a JSON file which isn't what we need. I think you need to set the accept header to `accept: application/xml` ? 

<img width="698" alt="image" src="https://github.com/Flagsmith/flagsmith/assets/14089968/a719d00e-8c9f-4b3d-a99f-3c889e1beae5">
<img width="443" alt="image" src="https://github.com/Flagsmith/flagsmith/assets/14089968/043a6ab5-660c-490a-b697-853dbc494f88">
<img width="185" alt="image" src="https://github.com/Flagsmith/flagsmith/assets/14089968/7d020745-ce94-4e62-989a-712d2d2a5143">

novakzaballa commented:
> 1. The IdP metadata is quite unmanageable in an auto-expanding text input. Ideally we'd have a file upload option here, but I think we'd need a new endpoint for that?

@matthewelwell Could the file be loaded locally and automatically parsed as XML, then converted into text? Would that be acceptable?


> 2. The bottom of the modal is missing padding (see screenshots below)

Done
<img width="659" alt="Screenshot 2024-06-06 at 11 39 57 AM" src="https://github.com/Flagsmith/flagsmith/assets/41410593/c45bd1ab-b712-485e-ac6a-65713a67ca9b">


> 3. It's not obvious that the 'Name' input label has a tooltip, let's add an icon there to make it clear. We should probably add tooltips to the remaining fields (probably with the exception of the IdP metadata which is fairly self explanatory.

Done:

<img width="661" alt="Screenshot 2024-06-06 at 11 39 50 AM" src="https://github.com/Flagsmith/flagsmith/assets/41410593/cf9c85b2-e32d-46bc-83d3-abbd45d6110b">


> 4. The alignment on the table is quite odd.

Corrected

<img width="704" alt="Screenshot 2024-06-06 at 11 40 09 AM" src="https://github.com/Flagsmith/flagsmith/assets/41410593/f406bf63-33d1-4df7-8830-57c2131fd5f0">


> 5. The 'Edit Configuration' button should say 'Update Configuration'

Changed

> 6. The 'Download Service Provider Metadata' option doesn't work correctly - it downloads a JSON file which isn't what we need. I think you need to set the accept header to `accept: application/xml` ?

Now we download .XML files instead of .JSON.

matthewelwell commented:
>@matthewelwell Could the file be loaded locally and automatically parsed as XML, then converted into text? Would that be acceptable?

I'm not sure I fully understand this, but it sounds reasonable - yep. Could we also then add a 'download' option to re-download the IdP metadata instead of showing it directly in the text input?

novakzaballa commented:
> I'm not sure I fully understand this, but it sounds reasonable - yep. Could we also then add a 'download' option to re-download the IdP metadata instead of showing it directly in the text input?
<img width="640" alt="Screenshot 2024-06-07 at 10 13 23 AM" src="https://github.com/Flagsmith/flagsmith/assets/41410593/81259703-15af-424a-a159-a290072e5c9b">

matthewelwell commented:
I'm not seeing the 'Show More' UI option here: 

<img width="707" alt="image" src="https://github.com/Flagsmith/flagsmith/assets/14089968/8c53650b-df27-41d2-bdbc-21abd5385d49">

It seems like the first time you click on one of the SAML configurations after browsing to the SAML settings page, it doesn't collapse the field. If you close it and reopen, it does collapse the field correctly. 

I also added a minor commit [here](https://github.com/Flagsmith/flagsmith/pull/4055/commits/04f3bbc62d6b76b596bbaceabd576b3961cea707) to ensure the 'Download IDP Metadata' option shows as clickable. Previously, it was just showing the regular cursor.

matthewelwell commented on docs/docs/system-administration/authentication/01-SAML/index.md:
```diff
@@ -8,45 +8,24 @@ SAML authentication requires an [Enterprise subscription](https://flagsmith.com/
+In the UI, you will be able to configure the following fields.
 
-## Setup (self-hosted)
+**Name:** (**Requierd**) A short name for the organisation, used as the input when clicking "Single Sign-on" at login
```
```suggestion
**Name:** (**Required**) A short name for the organisation, used as the input when clicking "Single Sign-on" at login
```

novakzaballa replied:
Done

matthewelwell commented on docs/docs/system-administration/authentication/01-SAML/index.md:
```diff
@@ -8,45 +8,24 @@ SAML authentication requires an [Enterprise subscription](https://flagsmith.com/
 
-To enable SAML for your Flagsmith organisation in a self-hosted environment, you will need access the
-[Django admin interface](/deployment/configuration/django-admin).
+**Frontend URL**: (**Requierd**) This should be the base URL of the Flagsmith dashboard.
```
```suggestion
**Frontend URL**: (**Required**) This should be the base URL of the Flagsmith dashboard, e.g. https://app.flagsmith.com for SaaS. 
```

novakzaballa replied:
Done

matthewelwell commented on docs/docs/system-administration/authentication/01-SAML/index.md:
```diff
@@ -8,45 +8,24 @@ SAML authentication requires an [Enterprise subscription](https://flagsmith.com/
-Once your SAML configuration is created, you can download your Flagsmith service provider metadata by going back to the
-list of SAML configurations in the Django admin interface and clicking "Download" on the SAML configuration you just
-created.
+Once Flagsmith has configured your identity provider, you can download the service provider metadata XML document whit
```
```suggestion
Once Flagsmith you have configured your identity provider, you can download the service provider metadata XML document with
```

novakzaballa replied:
Done

matthewelwell reviewed (changes requested):
I've tested the UI aspects and all seems good now. 

Added a few comments for some typos / improvements to the documentation.

matthewelwell commented on docs/docs/system-administration/authentication/01-SAML/index.md:
```diff
@@ -8,45 +8,24 @@ SAML authentication requires an [Enterprise subscription](https://flagsmith.com/
-Once your SAML configuration is created, you can download your Flagsmith service provider metadata by going back to the
-list of SAML configurations in the Django admin interface and clicking "Download" on the SAML configuration you just
-created.
+Once Flagsmith you have configured your identity provider, you can download the service provider metadata XML document
```
```suggestion
Once you have configured your identity provider, you can download the service provider metadata XML document
```

matthewelwell reviewed (approved):
Approved with one more minor typo (which I'll fix now)
